### DISCUSSION TITLE
Drag and drop and RequestPopupModelAction

### DISCUSSION THREAD

**Discussion by csamarchi** [eclipse-glsp/glsp#725] [General]
Hello,

I'm working with Node server and VS Code extension. I got part of my source model to render in the client, however, everything is stacked on top of each other in the left corner. I am able to click and drag, except when I let go of the click to drop, the module resumes back to its original spot stacked on top of the others.

I see this error in the server: 
![Screen Shot 2022-08-09 at 12 58 14 PM](https://user-images.githubusercontent.com/42501494/183725972-987425b5-fd88-44d9-863c-509ac613af2d.png)

What is the purpose of the `saveSourceModel()` function in the storage.ts file? Does this have anything to do with dragging and dropping and saving the position of where it's dropped? 

I'm surprised my modules are not laid out in the same format as the out of the box example, any ideas as to why? Is this something within the server or the client?

Maybe because edges are not defined properly? Here is my gmodel-factory:
```
 export class ModuleGModelFactory implements GModelFactory {
     @inject(ModuleModelState)
     protected modelState: ModuleModelState;

     createNodes(): void {
         // def? this.modelState.source.def
         const moduleArray: Module[] = this.modelState.source.def.modules.map((module: Module) => {
             const inputsArray = module.inputs.map((input: Input) => {
                 const tempBuild: Input = Input.builder().name(input.name).value(input.value).build();
                 return tempBuild;
             });
             const outputsArray = module.outputs.map((output: Output) => {
                 const tempBuild: Output = Output.builder().name(output.name).metaType(output.metaType).build();
                 return tempBuild;
             });
             const source = Source.builder().sourceType(module.source.sourceType).build();
             // eslint-disable-next-line max-len
             const solutionModule = Module.builder().name(module.name).source(source).inputs(inputsArray).outputs(outputsArray).build();
             return solutionModule;
         });
         this.modelState.modules = moduleArray;
     }

     createModel(): void {
         this.createNodes();
         const modules = this.modelState.modules;
         this.modelState.index.indexModule(modules);
         const moduleNodes = modules.map(module => this.createModule(module));
         const newRoot = GGraph.builder().id(module.id).addChildren(moduleNodes).revision(0).build();
         this.modelState.updateRoot(newRoot);
     }

     protected createModule(module: Module): GNode {
         const builder = GNode.builder() //
             .id(module.id)
             .addCssClass('tasklist-node')
             .add(GLabel.builder().text(module.name).id(`${module.id}_label`).build())
             .layout('hbox')
             .addLayoutOption('paddingLeft', 5)
             .position(module.position);

         if (module.size) {
             builder.addLayoutOptions({ prefWidth: module.size.width, prefHeight: module.size.height });
         }

         return builder.build();
     }
 }
```

**Comment by planger** [comment]
Hi @csamarchi,

The error you are seeing about the `RequestPopupModelAction` is unrelated to the positioning issue of your nodes. This seems to be because the [`RequestPopupModelActionHandler`](https://github.com/eclipse-glsp/glsp-server-node/blob/3996af680ef2904f6b7ac30ab7d21145d95f55d7/packages/server-node/src/features/popup/request-popup-model-action-handler.ts#L44) can't find the element for which the popup (hover widget) was requested. This is likely because the index isn't populated correctly for your elements?

Your `ChangeBoundsHandler` is responsible to process element moves and resizes invoked by the user, e.g. when they drag and drop an node to a new position on the client. As can be seen in the [change bounds handler in the project template](https://github.com/eclipse-glsp/glsp-examples/blob/master/project-templates/node-json-theia/glsp-server/src/handler/tasklist-change-bounds-handler.ts), it sets the position and size from a `ChangeBoundsOperation`, sent by the client after e.g. a move, in the source model. This should trigger a model update so that your `GModelFactory` is invoked again and hence it will create a new GModel with the new positions, as it reads the positions from the respective properties of the source model.

Unfortunately, I can't tell what's exactly wrong. It looks like you are reading positions from your source model and apply them to the GModel nodes in your `GModelFactory` here:
```
.position(module.position);
```

So I'd debug into that line and see what it actually writes into the position of the GModel nodes.

Your `ChangeBoundsHandler` in turn should obtain the correct `module` object from the GModel node ID (usually with the help of the index) and write the new position of the `ChangeBoundsOperation` back into `module.position` of the `modelState`. This way, it should get updated in the source model of the `modelState`, and thus considered on re-generation of the GModel on the next GModel update.

Since your module nodes seem to pop back to their original positions after a drag and drop, it looks like there is something missing in the process of reading the positions from the source model, writing them into the GModel and updating the source model after a `ChangeBoundsOperation`. So I'd step through the code in debug mode along those lines.

The purpose of `saveSourceModel()` is to store back the current version of your source model -- after potential modifications -- from your `modelState` into the persistent storage, such as a file, from where your source model originally has been loaded. It is usually invoked when the user saves the diagram (e.g. Ctrl-s) on a dirty diagram editor. See also https://www.eclipse.org/glsp/documentation/sourcemodel.

**Reply by csamarchi** [reply]
> This seems to be because the [`RequestPopupModelActionHandler`](https://github.com/eclipse-glsp/glsp-server-node/blob/3996af680ef2904f6b7ac30ab7d21145d95f55d7/packages/server-node/src/features/popup/request-popup-model-action-handler.ts#L44) can't find the element for which the popup (hover widget) was requested. This is likely because the index isn't populated correctly for your elements?

When you say index are you talking about the index file and id here? https://github.com/eclipse-glsp/glsp-examples/blob/master/project-templates/node-json-vscode/glsp-server/src/model/tasklist-model-index.ts

**Reply by csamarchi** [reply]
Where is the id suppose to be populated from? Right now, my id is a file path but I don't think that's right.

**Reply by csamarchi** [reply]
```
     createModel(): void {
         this.createNodes();
         const modules = this.modelState.modules;
         this.modelState.index.indexModule(modules);
         const moduleNodes = modules.map(module => this.createModule(module));
         const newRoot = GGraph.builder().id(module.id).addChildren(moduleNodes).revision(0).build();
         this.modelState.updateRoot(newRoot);
     }
```

My `module.id` in this function is currently a file path to my sourceUri, and not an id. I'm not seeing any other id's except those on the individual nodes? What am I missing?

**Reply by planger** [reply]
Yes, I meant the model index which usually maps two things:
1. GModel element IDs to GModel element instances in the model state
2. elements of the source model to elements of the generated GModel so that the server can identify the corresponding source model elements to be modified or queried based on the index, when actions from the client (which are based on the GModel element IDs, as the client doesn't know anything else) are sent to the server.

E.g. in the task list example:
https://github.com/eclipse-glsp/glsp-examples/blob/master/project-templates/node-json-vscode/glsp-server/src/model/tasklist-model-index.ts

As the server creates the GModel from the source model in the `GModelFactory` the index is also built during this step in the form of an source-model-ID to gmodel-element-ID map. In the task list example this is done here:
https://github.com/eclipse-glsp/glsp-examples/blob/00185505c83bef731c5a9435d01ced3d896d95f2/project-templates/node-json-vscode/glsp-server/src/model/tasklist-gmodel-factory.ts#L28

How you create an ID for the GModel element is up to you. Basically you just need to ensure that when you receive an action from the client, you can obtain from the GModel element ID the corresponding source model element ID, so that you can e.g. update it, etc.

From the error you posted above, I just figured that likely this line in the [RequestPopupModelActionHandler](https://github.com/eclipse-glsp/glsp-server-node/blob/3996af680ef2904f6b7ac30ab7d21145d95f55d7/packages/server-node/src/features/popup/request-popup-model-action-handler.ts#L44)
```typescript
const hoverElement = this.modelState.index.find(action.elementId);
```
doesn't find an element in your index for the `action.elementId`, which is I think the GModel root element ID.
So it seems you didn't put an entry for your GModel root element ID to the GModel root element into the map.

**Reply by csamarchi** [reply]
Where does the initial root element ID get set?
Is it in the gmodel-factory on this line:
```
const newRoot = GGraph.builder().id(module.id).addChildren(modules).revision(0).build();
```

I'm trying to figure out if the issue lies in the fact that my `module.id` is incorrect. If that's the case, I think I may need to restructure my source model. Right now I am creating `Module[]` from my source model, which is an array of modules. I think it should be something like `ModuleList` where `Module[]` is embedded within so that I can grab the id of the parent `ModuleList`.

**Reply by csamarchi** [reply]
One more question, correct me if I have this wrong..

Once we make a change, the client sends an action. Does the client _also_ send a new ID back to the server?

**Reply by planger** [reply]
Yes, that's the line where the root element ID is set in the `GModelFactory`.

> Once we make a change, the client sends an action.

Yes, it will send an operation, depending on the change. E.g. a `ChangeBoundsOperation` if the user moved or resized an element.
The client itself doesn't really make a change though. It just sends the aforementioned operation to request the server to perform the change, which then typically sends back an `UpdateModel` with the new GModel after the change is applied.

> Does the client also send a new ID back to the server?

No. The server is the only one who maintains and assigns IDs in the `GModelFactory`. The client just uses those IDs to refer to elements as a context of operations or actions.

**Reply by csamarchi** [reply]
Ok, does the server assign new id's to the same element on changes? Because I am seeing new id's after changes

**Reply by csamarchi** [reply]
> ```ts
> const hoverElement = this.modelState.index.find(action.elementId);
> ```
> 
> doesn't find an element in your index for the `action.elementId`, which is I think the GModel root element ID. So it seems you didn't put an entry for your GModel root element ID to the GModel root element into the map.

Could you elaborate on this? It's looking for the element ID in the index. What do you mean by entry for my GModel root element ID into the map?

**Reply by csamarchi** [reply]
Where in your example is taskList's id initially being set? 
This [line:](https://github.com/eclipse-glsp/glsp-examples/blob/00185505c83bef731c5a9435d01ced3d896d95f2/project-templates/node-json-vscode/glsp-server/src/model/tasklist-gmodel-factory.ts#L29) 
`        const newRoot= GGraph.builder().id(taskList.id).addChildren(childNodes).build();
`

I understand is setting the initial element root ID, but it's pulling `taskList.id` and I'm curious where this comes from. Is it the `builder()` function when we create the modules?

**Reply by planger** [reply]
> Ok, does the server assign new id's to the same element on changes? Because I am seeing new id's after changes

This depends on your implementation of the `GModelFactory`. That is, this is up to you. I would recommend to keep the GModel IDs stable to avoid flickering on updating the model. The framework itself doesn't generate new IDs. Your specific GModelFactory implementation is the only place where IDs should be generated and assigned to GModel elements (and then update the index).

> Could you elaborate on this? It's looking for the element ID in the index. What do you mean by entry for my GModel root element ID into the map?

The default implementation [`GModelIndex`](https://github.com/eclipse-glsp/glsp-server-node/blob/main/packages/server-node/src/features/model/gmodel-index.ts) keeps a map of ids to GModel elements. It looks like it can't find `action.elementId` in this map. When you call `modelState.updateRoot(yourGmodelRoot)`, the default implementation should invoke [`modelState.index.indexRoot(yourRoot)`](https://github.com/eclipse-glsp/glsp-server-node/blob/a06f0ae2372332225d0a9668eb3d3bf9b136587d/packages/server-node/src/features/model/model-state.ts#L97) to populate the index on each model update. Is this called in your scenario?
You can debug into that code and see what it actually indexes and which IDs are there and compare this to the ID that the `RequestPopupModelActionHandler` is trying to find in the index.

> I understand is setting the initial element root ID, but it's pulling taskList.id and I'm curious where this comes from. Is it the builder() function when we create the modules?

`taskList` is the source model of this example, so it is a custom domain-specific model that is represented by the diagram of your diagram editor. Thus `taskList.id` is just an id of this source model. Of course, task list is just a random example model. In your case this will likely be something different, depending on the domain of your diagram. Whether your domain model has a stable ID or not, or whether you somehow generate an ID, is up to you.
I however would recommend adding an ID to your source model and use it in your `GModelFactory` for the IDs of GModel elements that represent your source model element. This makes things more clear and simpler.

I hope this helps!

PS: Sorry, I needed to update this answer a bit, as I realized it isn't very clear in certain places.

**Comment by csamarchi** [comment]
Thank you, could you explain what the purpose of the index file is?

**Reply by planger** [reply]
It is used to export certain modules. E.g. if in package `my-package` there is a `src/abc/def/index.ts` that exports a file `mymodule.ts` which exports e.g. a class `MyClass`, other packages can add a dependency to `my-package` and import the exported classes, functions, etc. from this package by importing from `import { MyClass} from 'my-package/lib/abc/def'`. Otherwise it would have to be imported from the file directly `'my-package/lib/abc/def/mymodule'`. Thus if the filename changes or `MyClass` would be moved, everyone would have to  update their imports. With an index.ts the maintainers can just change it there and all imports can stay the same.

Note that this description simplifies it a lot as there are configurations in play like out folders (`lib`), etc. But this is roughly how it would work in the configuration of GLSP.

**Comment by csamarchi** [comment]
Thanks for the explanation.

I noticed in the `example.wf` file the initial positions radius's are given for each child like so:
```
      "children": [
        {
          "cssClasses": [],
          "type": "comp:header",
          "id": "task0_header",
          "position": {
            "x": 5,
            "y": 5
          },
          "size": {
            "width": 88.5859375,
            "height": 42
          },
```

Which is where the initial positioning is coming. Since we are not using the example file, we are using our own, how are we suppose to apply each position to each child? I don't expect us to hard code the initial x,y coordinates for each node, but I don't know how else?

**Reply by planger** [reply]
As you are using your own source models, the positions for each node need to come from your source models too. You have to store the positions in your source model or derive it from other properties of your source model or compute them with auto-layouting (e.g. with ELK) and eventually apply them to the GModel nodes you create in the GModelFactory.

GLSP doesn't put any restrictions or makes any assumptions on where you store the positions or how you compute them. Eventually however they need to end up in the GModel. It is really up to you where to get those positions from. Often they are just stored in the source model(s) alongside each object that end up as a node in the diagram.
If you don't want to "pollute" your original source model with this layout data though (e.g. if this format is not under your control), a common approach is also to have a second source model that just stores the positions (like this is x and y for that node in this other model).
